### ISSUE TITLE
Maybe unsound in SkBuff

### ISSUE THREAD

**Issue by lwz23** [aya-rs/aya#1112]
hello, thank you for your contribution in this project, I am scanning the unsoundness problem in rust project.
I notice the following code:
```
pub struct SkBuff {
    pub skb: *mut __sk_buff,
}


impl SkBuff {
    pub fn new(skb: *mut __sk_buff) -> SkBuff {
        SkBuff { skb }
    }

    #[allow(clippy::len_without_is_empty)]
    #[inline]
    pub fn len(&self) -> u32 {
        unsafe { (*self.skb).len }
    }
............................
}
```
Considering that `programs` is a pub mod and `skb` is a pub field, I assume that users can directly manipulate this field, and that `len` is a public function. This potential situation could result in `self.skb` being a null pointer, and directly dereferencing it might trigger undefined behavior (UB). For safety reasons, I felt it necessary to report this issue. If you have performed checks elsewhere that ensure this is safe, please don’t take offense at my raising this issue.
Sorry, I can't provide a Poc, because trying to import SkBuff in my environment will get an error:
```
error[E0432]: unresolved import `aya::programs::SkBuff`
 --> src/main.rs:3:5
  |
3 | use aya::programs::SkBuff;
  |     ^^^^^^^^^^^^^^^^^^^^^ no `SkBuff` in `programs`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `lwz` (bin "lwz") due to 1 previous error
```
If this can be solved, I am happy to provide a Poc.

**Comment by lwz23** [comment]
same problem for https://github.com/aya-rs/aya/blob/5a43bedc0180ba41854e6c23f0476c2cbeb1e2bd/ebpf/aya-ebpf/src/programs/sk_msg.rs#L9

**Comment by lwz23** [comment]
and maybe same problem for 
```
pub struct ProbeContext {
    pub regs: *mut pt_regs,
}

impl ProbeContext {
    pub fn new(ctx: *mut c_void) -> ProbeContext {
        ProbeContext {
            regs: ctx as *mut pt_regs,
        }
    }
    pub fn arg<T: FromPtRegs>(&self, n: usize) -> Option<T> {
        T::from_argument(unsafe { &*self.regs }, n)
    }
}
```

**Comment by lwz23** [comment]
If we need fix it maybe we should add check for null pointer in `new` method too :)

**Comment by vadorovsky** [comment]
I think the proper fix would be to use `NonNull` in context types and making sure it's a private field. We could provide `as_ptr() -> *mut pt_regs` (or whatever pointer type is wrapped by the given context) method for callers who want to do unsafe stuff with the context, or call some BPF helper directly.

**Comment by vadorovsky** [comment]
A pull request is very welcome! :slightly_smiling_face: Your build error looks weird, but hard to debug for me, since it's your own project where you're using Aya as an external crate. Would you mind sharing the code, or at least `Cargo.toml`?

That said, I think our integration tests should be good enough for you to make and test the change just using the main Aya repo.

**Comment by lwz23** [comment]
ofcourse, here is my main.rs:
```
extern crate aya;

use aya::programs::SkBuff;
fn main() {
    
}
```
and Cargo.toml:
```
[package]
name = "lwz"
version = "0.1.0"
authors = ["lwz"]
edition = "2018"

[dependencies]
aya = "0.13.1"
```
result:
```
error[E0432]: unresolved import `aya::programs::SkBuff`
 --> src/main.rs:3:5
  |
3 | use aya::programs::SkBuff;
  |     ^^^^^^^^^^^^^^^^^^^^^ no `SkBuff` in `programs`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `lwz` (bin "lwz") due to 1 previous error
lwz@LAPTOP-F68E79VH:/mnt/e/Github/lwz$
```

**Comment by lwz23** [comment]
and https://github.com/aya-rs/aya/blob/1dfcfbcf90d4b01ec56c4bc236577857c1fb6753/ebpf/aya-ebpf/src/programs/sk_buff.rs#L263
I'm not sure if I've found all of these bugs, but it seems to be a common problem in this library. Maybe my program is underreporting. But that's probably all the cases I could find.

**Comment by vadorovsky** [comment]
Yeah, if you find any other structs wrapping a raw pointer, it totally makes sense to ensure it's `NonNull` and private.

**Comment by lwz23** [comment]
Considering this is a unsound problem and this crate is published on crates.io, I suggest maybe we should report it to RustSec?

**Comment by ajwerner** [comment]
> Considering this is a unsound problem and this crate is published on crates.io, I suggest maybe we should report it to RustSec?

I doubt it. This crate (aya-ebpf) can only be compiled with nightly compiler targeting bpfeb-unknown-none and bpfel-unknown-none, which are [Tier 3](https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-3) platforms.
